hybrid/pkg/server/server_interceptor.go

```diff
@@ -0,0 +1,113 @@
+package hybridserver
+
+import (
+	"context"
+
+	hclog "github.com/hashicorp/go-hclog"
+	nodeattestorv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/server/nodeattestor/v1"
+)
+
+type ServerInterceptor interface {
+	Recv() (*nodeattestorv1.AttestRequest, error)
+	Send(resp *nodeattestorv1.AttestResponse) error
+	SetContext(ctx context.Context)
+	Context() context.Context
+	SetLogger(logger hclog.Logger)
+	SetReq(req *nodeattestorv1.AttestRequest)
+	CanReattest() []bool
+	SpiffeID() string
+	SetSpiffeID(spiffeID string)
+	CombinedSelectors() []string
+	Stream() nodeattestorv1.NodeAttestor_AttestServer
+	NewInterceptor() ServerInterceptor
+	setCustomStream(stream nodeattestorv1.NodeAttestor_AttestServer)
+}
+
+type HybridPluginServerInterceptor struct {
+	nodeattestorv1.NodeAttestor_AttestServer
+
+	stream            nodeattestorv1.NodeAttestor_AttestServer
+	logger            hclog.Logger
+	ctx               context.Context
+	req               *nodeattestorv1.AttestRequest
+	response          *nodeattestorv1.AttestResponse
+	combinedSelectors []string
+	spiffeID          string
+	canReattest       []bool
+}
+
+func (m *HybridPluginServerInterceptor) NewInterceptor() ServerInterceptor {
+	return &HybridPluginServerInterceptor{
+		ctx:               m.ctx,
+		stream:            m.stream,
+		logger:            m.logger,
+		req:               m.req,
+		response:          m.response,
+		combinedSelectors: m.combinedSelectors,
+		spiffeID:          m.spiffeID,
+		canReattest:       m.canReattest,
+	}
+}
+
+func (m *HybridPluginServerInterceptor) Recv() (*nodeattestorv1.AttestRequest, error) {
+	return m.req, nil
+}
+
+func (m *HybridPluginServerInterceptor) Send(resp *nodeattestorv1.AttestResponse) error {
+	if x, ok := resp.GetResponse().(*nodeattestorv1.AttestResponse_AgentAttributes); ok {
+
+		m.combinedSelectors = append(m.combinedSelectors, x.AgentAttributes.SelectorValues...)
+		if len(m.spiffeID) == 0 {
+			m.spiffeID = x.AgentAttributes.SpiffeId
+		}
+
+		m.canReattest = append(m.canReattest, x.AgentAttributes.CanReattest)
+	}
+
+	return nil
+}
+
+func (m *HybridPluginServerInterceptor) SetContext(ctx context.Context) {
+	m.ctx = ctx
+}
+
+func (m *HybridPluginServerInterceptor) Context() context.Context {
+	return m.ctx
+}
+
+func (m *HybridPluginServerInterceptor) SetLogger(logger hclog.Logger) {
+	m.logger = logger
+}
+
+func (m *HybridPluginServerInterceptor) SetReq(req *nodeattestorv1.AttestRequest) {
+	m.req = req
+}
+
+func (m *HybridPluginServerInterceptor) CanReattest() []bool {
+	return m.canReattest
+}
+
+func (m *HybridPluginServerInterceptor) SpiffeID() string {
+	return m.spiffeID
+}
+
+func (m *HybridPluginServerInterceptor) CombinedSelectors() []string {
+	return m.combinedSelectors
+}
+
+func (m *HybridPluginServerInterceptor) Stream() nodeattestorv1.NodeAttestor_AttestServer {
+	return m.stream
+}
+
+func (m *HybridPluginServerInterceptor) SetSpiffeID(spiffeID string) {
```

Comment:
This is not used, i'll remove it. I was using that, but removed and I'm doing this now:

if len(m.spiffeID) == 0 {
			m.spiffeID = x.AgentAttributes.SpiffeId
		}
		
And no, it does not goes againt the logic of the spiffeid, because it's determined in the server.go, in the attest method

Reply:
this is a test used in the same package? or another one? 
generally this kind of functions are not exposed, because it can allow changing a calculated value without validations